tasks/src/lib.rs

```diff
@@ -0,0 +1,337 @@
+
+            match event {
+                ClientFlowEvent::CommandSent { tag, handle } => {
+                    // This `unwrap` can't fail because `waiting_tasks` contains all unsent `Commands`.
+                    let entry = self.waiting_tasks.remove(&handle).unwrap();
+                    self.active_tasks.insert(handle, (tag, entry));
+                }
+                ClientFlowEvent::CommandRejected { handle, status, .. } => {
+                    let body = match StatusV2::from(status) {
+                        StatusV2::Tagged(Tagged { body, .. }) => body,
+                        _ => unreachable!(),
+                    };
+
+                    let (_, task) = self.active_tasks.remove(&handle).unwrap();
+
+                    let output = Some(task.process_tagged(body));
+
+                    return Ok(SchedulerEvent::TaskFinished(TaskToken { handle, output }));
+                }
+                ClientFlowEvent::DataReceived { data } => {
+                    if let Some(data) = trickle_down(data, self.active_tasks_mut(), |task, data| {
+                        task.process_data(data)
+                    }) {
+                        return Ok(SchedulerEvent::Unsolicited(Response::Data(data)));
+                    }
+                }
+                ClientFlowEvent::StatusReceived { status } => match StatusV2::from(status) {
+                    StatusV2::Untagged(body) => {
+                        if let Some(body) =
+                            trickle_down(body, self.active_tasks_mut(), |task, body| {
+                                task.process_untagged(body)
+                            })
+                        {
+                            return Ok(SchedulerEvent::Unsolicited(Response::Status(
+                                Status::from(StatusV2::Untagged(body)),
+                            )));
+                        }
+                    }
+                    StatusV2::Bye(bye) => {
+                        if let Some(bye) =
+                            trickle_down(bye, self.active_tasks_mut(), |task, bye| {
+                                task.process_bye(bye)
+                            })
+                        {
+                            return Ok(SchedulerEvent::Unsolicited(Response::Status(
+                                Status::from(StatusV2::Bye(bye)),
+                            )));
+                        }
+                    }
+                    StatusV2::Tagged(Tagged { tag, body }) => {
+                        let handle = {
+                            if let Some((handle, _)) =
+                                self.active_tasks.iter().find(|(_, (tag_, _))| tag == *tag_)
+                            {
+                                *handle
+                            } else {
+                                return Err(SchedulerError::FatalProtocolIssue(
+                                    FatalProtocolIssue::UnexpectedTaggedResponse,
```

Comment:
`UnexpectedTaggedResponse` should contain the response. Then the caller is able to log it.